src/relcast.erl

```diff
@@ -1050,11 +1072,112 @@ make_seq(ID, #state{seq_map=SeqMap, ids=A}=State) ->
+    KeyTree = do_serialize(Mod, Old, New, ?stored_key_prefix, Transaction),
+    S1 = maybe_write_key_tree(KeyTree, S),
+    S1#state{old_serialized = New, transaction_dirty = true}.
+
+maybe_write_key_tree(KeyTree, S) when KeyTree == S#state.key_tree->
+    S;
+maybe_write_key_tree(KeyTree, S) ->
+    ok = rocksdb:transaction_put(S#state.transaction, ?stored_key_tree,
+                                 term_to_binary(KeyTree, [compressed])),
+    S#state{key_tree = KeyTree}.
+
+old_size(M) when is_map(M) ->
+    maps:size(M);
+old_size(_) ->
+    -1.
+
+do_serialize(Mod, Old, New, Prefix, Transaction) ->
+    case New of
+        State when is_binary(State) ->
+            ok = rocksdb:transaction_put(Transaction, ?stored_module_state, State),
+            bin;
+        StateMap ->
+            S = lists:sort(maps:to_list(StateMap)),
+            SSize = maps:size(StateMap),
+            OSize = old_size(Old),
+            O = case Old of
+                    %% since we're serializing for the first time, we need to make sure that
+                    %% everything gets written out, otherwise we have partial state that
+                    %% won't restore correctly.
+                    BorU when BorU == undefined orelse
+                              is_binary(BorU) ->
+                        lists:map(fun({K, _V}) -> {K, never_ever_match_with_anything} end, S);
+                    _Size when OSize =/= SSize ->
+                        SKeys = maps:keys(StateMap),
+                        OKeys = maps:keys(Old),
+                        Old1 = lists:foldl(fun(L, OM) ->
+                                                   OM#{L => undefined}
+                                           end,
+                                           Old,
+                                           SKeys -- OKeys),
+                        maps:to_list(maps:without(OKeys -- SKeys, Old1));
+                    _ ->
+                        lists:sort(maps:to_list(Old))
+                end,
+            L = lists:zip(S, O),
+            KeyTree =
+                lists:map(
+                  fun({{K, V}, {_, V}}) ->
+                          %% should be a binary
+                          K;
+                     ({{K, V}, {_, OV}}) ->
+                          KeyName = <<Prefix/binary, (atom_to_binary(K, utf8))/binary>>,
+                          case is_map(V) of
+                              true ->
+                                  do_serialize(K, fixup_old_map(OV), V, <<KeyName/binary, "_">>, Transaction);
+                              false ->
+                                  ok = rocksdb:transaction_put(Transaction, KeyName, V),
+                                  K
+                          end
+                  end,
+                  L),
+            [Mod | KeyTree]
+    end.
+
+fixup_old_map(never_ever_match_with_anything) ->
+    undefined;
+fixup_old_map(M) ->
+    M.
+
+%% this needs to return {serstate, restored modstate} I think
```

Comment:
So should it?

Reply:
I'll clean up all the todos and stray comments.